CLEVA Scenarios, Perturbations and Metrics

lyy1994 opened the pull request:
# Summary

This PR implements major functionalities of [CLEVA](https://arxiv.org/pdf/2308.04813.pdf):
- All Chinese scenarios with a multiple prompt templates setting.
- All Chinese perturbation strategies.
- Metrics tailored to Chinese evaluation (e.g., Chinese word segmentation, Chinese word lists, etc.).

# Command

An example of running a CLEVA scenario (Chinese-to-English translation) is as follows (We need to specify `prompt_id` as CLEVA provides multiple prompt templates; It starts from 0 and is 0 by default):
```sh
helm-run \
-r "cleva:model=openai/gpt-3.5-turbo-0613,task=translation,subtask=zh2en,prompt_id=0,version=v1,data_augmentation=cleva" \
--num-train-trials <num_trials> \
--max-eval-instances <max_eval_instances> \
--suite <suite_id>
```

# Results

Below is the result comparison of some CLEVA scenarios using `openai/gpt-3.5-turbo-0613`:

| Scenario | Metric | Reproduced | CLEVA |
| ---- | ----------------- | ---------------- | ----------- |
| task=summarization,subtask=dialogue_summarization | ROUGE-2 | 0.3045 | 0.3065 |
| task=translation,subtask=en2zh | SacreBLEU | 60.48 | 59.23 |
| task=fact_checking | Exact Match | 0.4595 | 0.4528 |
| task=bias,subtask=dialogue_region_bias | Micro F1 | 0.5656 | 0.5589 |

yifanmai commented on src/helm/benchmark/augmentations/cleva_perturbation.py:
```diff
@@ -0,0 +1,722 @@
+    @property
+    def description(self) -> PerturbationDescription:
+        """Return a perturbation description for this class."""
+        source_str = ",".join([f"{k}={v}" for k, v in self.source_class.items()])
+        target_str = ",".join([f"{k}={v}" for k, v in self.target_class.items()])
```
Why is this necessary as opposed to just passing in `self.source_class` and `self.target_class`?

lyy1994 replied:
I think either way works: here we just follow the dict representation in [person name perturbation](https://github.com/stanford-crfm/helm/blob/65fdf7a838e06d07edd21ca408f9c56717ea90cb/src/helm/benchmark/augmentations/person_name_perturbation.py#L189).

yifanmai commented on src/helm/benchmark/augmentations/cleva_perturbation.py:
```diff
@@ -0,0 +1,722 @@
+
+            words_and_similar_word_dict["similar_pinyin_words"] = similar_word_pinyin_list
+            words_and_similar_word_dict_list.append(words_and_similar_word_dict)
+        return words_and_similar_word_dict_list
```
Instead of a structure that's like

```
[
  {
    "original_word": "word1"
    "similar_pinyin_words": ["word2", "word3"]
  }
]
```

could you just do:

```
{
  "word1": ["word2", "word3"]
}
```

yifanmai commented on src/helm/benchmark/augmentations/cleva_perturbation.py:
```diff
@@ -0,0 +1,722 @@
+                self.consider_tone,
+                rng,
+            )
+            for dict in words_and_similar_word_dict_list:
```
Don't name a variable `dict` because that is a name of a Python built-in.

If you follow my later suggestion in `get_words_with_similar_pinyin`, then this just becomes

```
for original_word, similar_pinyin_words in words_and_similar_word_dict.items():
```

yifanmai reviewed (changes requested):
Thanks, this is a very high quality PR!

I'm sending you a partial review first of the metrics and perturbations. It will take me a little longer to finish reviewing the rest of the PR because it is rather long.

lyy1994 commented on src/helm/benchmark/metrics/classification_metrics.py:
```diff
@@ -68,3 +70,38 @@ def evaluate_instances(self, request_states: List[RequestState]) -> List[Stat]:
         ]
+
+
+class CLEVAMultipleChoiceClassificationMetric(Metric):
```
Yes, you are right.

yifanmai replied:
Sounds good. I think you can just drop `CLEVA` from the name since this can be used for the non-CLEVA scenarios as well.

Likewise for "CLEVA" in the remainder of the comments and function body.

lyy1994 replied:
Thanks for your suggestions! We drop `CLEVA` in the latest commits.

lyy1994 commented on src/helm/benchmark/metrics/cleva_accuracy_metrics.py:
```diff
@@ -0,0 +1,54 @@
+
+    It accepts multiple predictions and multiple references to calculate the accuracy
+    per instance. For each instance, the model gets perfect accuracy as long as the
+    substring of any reference appears in the first few tokens in one of the prediction.
```
You are right. This precision metric is designed for a special scenario, where substrings of the given reference are very likely to be another correct answer. You can refer to [this paper](https://openreview.net/pdf?id=gJcEM8sxHK) for more details.

lyy1994 commented on src/helm/benchmark/metrics/basic_metrics.py:
```diff
@@ -264,6 +264,34 @@ def bleu_1(gold: str, pred: str) -> float:
+        def tokenize(self, text: str) -> List[str]:
+            return [char for char in text]
+
+    char_tokenizer = CharTokenizer()
```
Chinese tokenizers are mandatory for BLEU and ROUGE scores because (almost) no empty space is provided in the string such that the default tokenizer can correctly split the text. We follow the idea of [sacrebleu](https://github.com/mjpost/sacrebleu/blob/bafca9df5b933770aa5c7fb7d858acd34ac43d4c/sacrebleu/tokenizers/tokenizer_zh.py#L93) in `CLEVAMachineTranslationMetric` and implement a Chinese character tokenizer here. For ROUGE score, this tokenizer is utilized by passing it to the `tokenizer` argument of the `__init__` function of `rouge_scorer.RougeScorer`. For BLEU, we directly split the Chinese string into a list of characters before feeding it to `sentence_bleu`.

lyy1994 replied:
We made a bit of modification such that the Chinese tokenizer usage across metrics is clearer and unified.

lyy1994 commented:
> Thanks, this is a very high quality PR!
> 
> I'm sending you a partial review first of the metrics and perturbations. It will take me a little longer to finish reviewing the rest of the PR because it is rather long.

Thanks for your instructive comments! I believe we have addressed all the issues you mentioned in previous review comments and look forward to your kind review of the rest of our PR as well as the recent revision.

yifanmai reviewed (changes requested):
More review comments. I have only the run / scenario stuff left to review, which I should be able to do tomorrow.

lyy1994 commented on src/helm/benchmark/metrics/cleva_harms_metrics.py:
```diff
@@ -0,0 +1,253 @@
+
+    METHOD_LIST = ["char", "jieba"]
+
+    def __init__(self, method: str = "char") -> None:
```
"jieba" is actually not used in the metric part. We just removed this option.

lyy1994 commented:
> More review comments. I have only the run / scenario stuff left to review, which I should be able to do tomorrow.

Thank you for your prompt response. We have made the necessary adjustments based on your latest review. Once you release the comments about run / scenarios, we will immediately address any concerns you may have.

yifanmai commented on src/helm/benchmark/scenarios/cleva_scenario.py:
```diff
@@ -0,0 +1,1491 @@
+        if "choices" in data.keys():
+            # We take the corresonding choices and apply the `label` template
+            # Note: we do not allow `label` template to access other fields in multi-choice tasks
+            choices = [self._apply(c, templates.get("label", None), label=c) for c in data["choices"]]
```
Can this be moved to right before `data["choices"] = choices`? Then you don't need two instances of `if "choices" in data.keys():`

yifanmai commented on src/helm/benchmark/scenarios/cleva_scenario.py:
```diff
@@ -0,0 +1,1491 @@
+        for k, template in templates.items():
+            if k not in ["input", "verbalizer", "meta", "instruction", "label", "answer_context"]:
+                assert k in data, f"Template key `{k}` is not valid!"
+                transformed_data[k] = self._apply(data[k], template, **data)
```
If I'm not wrong, you should `continue` if `k` is "input" or "verbalizer" - "input" is only used in the final merge and operates on `transformed_data` rather than `data` (or can `input` recursively include itself?). And "verbalizer" was already used earlier.

Jianqiao-Zhao replied:
I am sorry that I don't think I fully understand your valuable suggestion. 

Below, I attach a template example. With the code you commented on here, we will first convert/organize the "choices" field into a string. Then, we will use all the information from the template and from the raw data to complete the "input" field.
```
{
    "choices": {
        "item_separator": "\n",
        "item_template": "{idx}. {item}",
        "item_index": "upper"
    },
    "instruction": "请根据文段内容补全下划线处的成语。",
    "input": "{content}\n",
    "answer_context": "答：（请用选项序号回答）",
    "meta": {
        "language": "zh-Hans",
        "mul_as_gen": true,
        "postprocess": [
            "common",
            "upper"
        ]
    }
}
```

yifanmai replied:
Sorry, I misread the code. I think what you did originally is correct.

I would suggest putting an example of "template" in the comments somewhere, so that it is easier to understand.

lyy1994 replied:
Thanks! We have added examples for both `_apply` and `_apply_all` methods to better illustrate the underlying workflow.

yifanmai commented on src/helm/benchmark/scenarios/cleva_scenario.py:
```diff
@@ -0,0 +1,1491 @@
+            # If data is a `Dict[str, str]`, flatten all its key-value pairs and treat them as additional fields
+            if isinstance(data, dict):
+                return template.format(**kwargs, **data)
+            # If data is a `str`, just directly compose the output string from other stringified fields
+            return template.format(**kwargs)
```
Is `data` unused in this branch? What's an example of when this happens?

lyy1994 replied:
Yes. I think the code comment is somewhat misleading and we have revised this. In our implementation, when `template` is a string and `data` is not a dict, it will compose a new string from fields in `kwargs`. `data` will not be used because it does not have a name associated, meaning that we have no idea where `data` should be placed when formatting string with `template`. But it does not mean that `data` could not be accessed by `template`, since `data` and its associated name will be available in `kwargs`. An example is the `input` field ("{review} It is"), which basically composes the prompt "This movie sucks. It is" from fields in `kwargs` (`{"review": "This movie sucks.", "label": [0], "choices": ["negative", "positive"]}`). Another example is the `label` field (" {label}" which adds space in front for language modeling), which does not use `data` but `label` in `kwargs` (`{"label": "How about you?"}`), yet they are the same.

yifanmai commented on src/helm/benchmark/scenarios/cleva_scenario.py:
```diff
@@ -0,0 +1,1491 @@
+            data["label"] = [self._apply(x, templates.get("label", None), **kwargs, label=x) for x in data["label"]]
+        return data
+
+    def _apply(self, data: Any, template: Optional[Any], **kwargs) -> str:
```
You should do something like:

```python
# At the module level
TemplateData = Union[str, Dict[str, str], List[str], List[Dict[str, str]]]
Template = Union[str, Dict[str, str]]

# Here
def _apply(self, data: TemplateData, template: Optional[Template], **kwargs) -> str:
```

Also update the docstring to indicate that template can be `None`.

lyy1994 replied:
Thanks for your suggestions! We have annotated types according to your advice, except that we use `RawData` instead of `TemplateData` as the data is expected not to be tied to a specific prompt template. We notice that the static analyzer cannot narrow `Union` down to a more specific type and raise errors. In this case, we use `# type: ignore` but add some assertions to guarantee the data type.

yifanmai commented on src/helm/benchmark/scenarios/cleva_scenario.py:
```diff
@@ -0,0 +1,1491 @@
+                        ]
+                    )
+            else:
+                raise ValueError(f"Unsupport input: {data}")
```
`Unsupport` -> `Unsupported`

yifanmai commented on src/helm/benchmark/scenarios/cleva_scenario.py:
```diff
@@ -0,0 +1,1491 @@
+        )
+        return instance
+
+    def _apply_all(self, data: dict, templates: dict) -> dict:
```
Sounds like templates should be of type `templates: Dict[str, Optional[Template]]` (see below for the definition of `Template`)

yifanmai commented on src/helm/benchmark/run_specs.py:
```diff
@@ -625,6 +652,114 @@ def get_instruction_following_critique_metric_specs(num_respondents: int) -> Lis
+        "dialogue_generation:task_oriented": partial(get_basic_metric_specs, ["chinese_bleu_1"]),
+        "data_to_text_generation": partial(get_basic_metric_specs, ["chinese_bleu_1"]),
+        "mathematical_reasoning:math_world_problem": partial(get_basic_metric_specs, ["cleva_math_result_match"]),
+    }
```
Could you create a new file `src/helm/benchmark/presentation/cleva_run_specs.conf` containing example run spec entries? This will help users discover what tasks and subtasks are available. This will alse cause GitHub actions to test that your run spec function correctly parses and run spec entries. You can look at [`biomedical_run_specs.conf`](https://github.com/stanford-crfm/helm/blob/v0.2.3/src/helm/benchmark/presentation/run_specs_biomedical.conf) for an example.

lyy1994 replied:
No problem. In the latest commits, we add `run_specs_cleva_v1.conf` (we add a `v1` tag for future compatibility), which lists all tasks / subtasks and `prompt_id` available.

yifanmai commented on src/helm/benchmark/run_specs.py:
```diff
@@ -2243,6 +2378,116 @@ def get_anthropic_hh_rlhf_spec(num_respondents: int, subset: str) -> RunSpec:
 
 
+@run_spec_function("cleva")
+def get_cleva_spec(task: str, version: str, subtask: str = None, prompt_id: int = 0) -> RunSpec:
```
Is there documentation somewhere about the valid options for `prompt_id`?

lyy1994 replied:
We list all available `prompt_id` in `run_specs_cleva_v1.conf`. This should help users find all valid `prompt_id` options.

yifanmai commented on src/helm/benchmark/run_specs.py:
```diff
@@ -2243,6 +2378,116 @@ def get_anthropic_hh_rlhf_spec(num_respondents: int, subset: str) -> RunSpec:
 
+@run_spec_function("cleva")
+def get_cleva_spec(task: str, version: str, subtask: str = None, prompt_id: int = 0) -> RunSpec:
+    CLEVAScenario.download_dataset(version)
```
Don't call `download_dataset` here. The downloading should happen inside `get_instances()` instead.

Jianqiao-Zhao replied:
`download_dataset` not only downloads training and testing data. It downloads prompt templates and inference parameters for each scenario/subtask as well. We need to obtain prompt templates and inference parameters during the construction of the `RunSpec`, before `get_instances` is called. 

Also because a scenario is not instantiated yet, we cannot really access the `self.output_path` when downloading prompt templates and inference parameters. We can download the data separately, but this solution may not be elegant either.

We are eager to learn your insights on solving this dilemma. We will take immediate action according to your guidance. Thank you!

yifanmai replied:
Thanks for the explanation. Let's merge this for now. I'll eventually open a new PR to provide another way of getting this path, so that we don't need to hardcode it.

yifanmai commented on src/helm/benchmark/scenarios/cleva_scenario.py:
```diff
@@ -0,0 +1,1491 @@
+        return "language_modeling"
+
+    def process_instance(self, row: Dict[str, Any], split: str) -> Instance:
+        text: str = row["choices"][0]  # The length of row["choices"] should be 1. Only the first one is used.
```
Add `assert(len(row["choices"]))`

yifanmai commented on src/helm/benchmark/run_specs.py:
```diff
@@ -2243,6 +2378,116 @@ def get_anthropic_hh_rlhf_spec(num_respondents: int, subset: str) -> RunSpec:
 
+@run_spec_function("cleva")
+def get_cleva_spec(task: str, version: str, subtask: str = None, prompt_id: int = 0) -> RunSpec:
+    CLEVAScenario.download_dataset(version)
```
Add `from .scenarios.cleva_scenario import CLEVAScenario  # noqa` as the first line of this method.

yifanmai commented on src/helm/benchmark/run_specs.py:
```diff
@@ -44,6 +45,7 @@
     TaskType,
     get_lextreme_task_type,
 )
+from .scenarios.cleva_scenario import CLEVAScenario
```
Remove this import and move it to inside `get_cleva_spec`. Otherwise `ImportError` will be raised for users who don't install the optional dependencies.

yifanmai commented on src/helm/benchmark/run_specs.py:
```diff
@@ -2243,6 +2378,116 @@ def get_anthropic_hh_rlhf_spec(num_respondents: int, subset: str) -> RunSpec:
+        scenario_spec=scenario_spec,
+        adapter_spec=adapter_spec,
+        metric_specs=metric_specs,
+        groups=["cleva"],
```
Do: `groups=["cleva", f"cleva_{task}"],`

yifanmai commented on src/helm/benchmark/run_specs.py:
```diff
@@ -2243,6 +2378,116 @@ def get_anthropic_hh_rlhf_spec(num_respondents: int, subset: str) -> RunSpec:
+        scenario_spec=scenario_spec,
+        adapter_spec=adapter_spec,
+        metric_specs=metric_specs,
+        groups=["cleva"],
```
Note that you will need to edit `schema.yaml` to add the new scenarios and metrics. I would recommend doing this in a new Pull Request rather than in this Pull Request, because editing `schema.yaml` is rather complicated.

Jianqiao-Zhao replied:
Thanks for pointing it out. We will edit `schema.yaml` in a new Pull Request.

yifanmai replied:
I've started a template for the schema.yaml in this PR #1845. I will get it merged soon, and then it would be great to have your help in checking it and filling it in.

yifanmai reviewed (changes requested):
Thanks for the revisions so far!

Most of this is good, but I need more time to take a closer look at the template code, because that is rather complicated.

yifanmai commented on src/helm/benchmark/scenarios/cleva_scenario.py:
```diff
@@ -0,0 +1,1530 @@
+        """
+
+        labels: List[str] = copy.deepcopy(row["label"])
+        instance = self.converter.transform(row, self.prompt_template, split)
+        if split == TEST_SPLIT:
+            instance = Instance(
+                input=instance.input,
+                references=[Reference(Output(text=label), tags=[CORRECT_TAG]) for label in labels],
+                split=split,
+            )
```
Minor suggestion:

```
        if split == TEST_SPLIT:
            return Instance(
                input=instance.input,
                references=[Reference(Output(text=label), tags=[CORRECT_TAG]) for label in labels],
                split=split,
            )
        else:
            return self.converter.transform(row, self.prompt_template, split)
```

This avoids making an unnecessary `transform` for instances in the test split.

lyy1994 replied:
`transform` is a must because we need to apply the template to get the correct prompt; we overwrite `references` here because `transform` will modify `references` to incorporate CoT, which is only desired for train instances but not for test instances. So we put the raw `references` back to ensure a correct evaluation.

yifanmai replied:
Understood, thanks.

yifanmai commented on src/helm/benchmark/scenarios/cleva_scenario.py:
```diff
@@ -0,0 +1,1530 @@
+        prompt_templates = cls.load_prompt_templates(task, subtask, version)
+        if prompt_id >= len(prompt_templates):
+            raise ValueError(
+                f"You want to use {prompt_id + 1}-th prompt template, but there is only"
```
I'd suggest changing `{prompt_id + 1}-th prompt template` to `prompt template with prompt_id {prompt_id}` for simplicity.

yifanmai commented on src/helm/benchmark/scenarios/cleva_scenario.py:
```diff
@@ -0,0 +1,1530 @@
+                # If the value is a list, then look into its entries
+                elif isinstance(data[_name], list):
+                    assert len(data[_name]) > 0, f"The length of {_name} must be larger than 0."
+                    if isinstance(data[_name][0], int):
```
This only maps the first element of the index; do you need to also map the remaining elements?

I'm also wondering if you can use recursion handle the inner elements.

lyy1994 replied:
Actually, we use the first element to check the type to determine how the mapping should apply. The mapping will be applied to every element in the list, as shown in the following couple of lines.

yifanmai reviewed (changes requested):
OK, I have finished reviewing. I was confused about the templating code before but I understand it better now.

yifanmai commented on src/helm/benchmark/presentation/run_specs_cleva_v1.conf:
```diff
@@ -0,0 +1,298 @@
+entries: [
+    ######################################################### Application ######################################################
+
+    {description: "cleva:model=full_functionality_text,task=dialogue_generation,subtask=task_oriented,prompt_id=0,version=v1,data_augmentation=cleva", priority: 1}
```
For runs that use the `ADAPT_MULTIPLE_CHOICE_JOINT` or `ADAPT_GENERATION` adapters, could you change `model=full_functionality_text` to `model=text`?

The reason for this is that some models like GPT-3.5 and GPT-4 are _not_ `full_functionality_text` models because they don't return logprobs, but we still want to be able to run them on a subset of these scenarios that don't need logprobs.

yifanmai reviewed (changes requested):
I tried running this on the debugging model `simple/model1`. Most runs work, but some are failing inside template rendering:


Failing runs:

```
cleva:task=dialogue_generation,version=v1,prompt_id=0,subtask=task_oriented,model=simple/model1,data_augmentation=cleva
cleva:task=paraphrase_identification,version=v1,prompt_id=0,subtask=short_utterance,model=simple/model1,data_augmentation=cleva
cleva:task=paraphrase_identification,version=v1,prompt_id=0,subtask=financial_question,model=simple/model1,data_augmentation=cleva
cleva:task=summarization,version=v1,prompt_id=0,subtask=dialogue_summarization,model=simple/model1,data_augmentation=cleva
cleva:task=closed_book_question_answering,version=v1,prompt_id=0,subtask=medical_question_answering,model=simple/model1,data_augmentation=cleva
cleva:task=text_classification,version=v1,prompt_id=0,subtask=news,model=simple/model1,data_augmentation=cleva
cleva:task=text_classification,version=v1,prompt_id=0,subtask=humor,model=simple/model1,data_augmentation=cleva
cleva:task=sentiment_analysis,version=v1,prompt_id=0,model=simple/model1,data_augmentation=cleva
cleva:task=coreference_resolution,version=v1,prompt_id=0,model=simple/model1,data_augmentation=cleva
cleva:task=commonsense_reasoning,version=v1,prompt_id=0,subtask=textual_entailment,model=simple/model1,data_augmentation=cleva
cleva:task=code_synthesis,version=v1,prompt_id=0,model=simple/model1,data_augmentation=cleva
cleva:task=toxicity_detection,version=v1,prompt_id=0,model=simple/model1,data_augmentation=cleva
cleva:task=bias,version=v1,prompt_id=0,subtask=dialogue_gender_bias,model=simple/model1,data_augmentation=cleva
cleva:task=bias,version=v1,prompt_id=0,subtask=dialogue_occupation_bias,model=simple/model1,data_augmentation=cleva
cleva:task=bias,version=v1,prompt_id=0,subtask=dialogue_race_bias,model=simple/model1,data_augmentation=cleva
cleva:task=bias,version=v1,prompt_id=0,subtask=dialogue_region_bias,model=simple/model1,data_augmentation=cleva
cleva:task=fact_checking,version=v1,prompt_id=0,model=simple/model1,data_augmentation=cleva
```

Error traceback:
```
Traceback (most recent call last):
  File "/home/yifanmai/oss/helm/src/helm/benchmark/runner.py", line 173, in run_all
    self.run_one(run_spec)
  File "/home/yifanmai/oss/helm/src/helm/benchmark/runner.py", line 221, in run_one
    instances = scenario.get_instances()
  File "/home/yifanmai/oss/helm/src/helm/benchmark/scenarios/cleva_scenario.py", line 942, in get_instances
    instances.extend(self.process_dialogue_instance(row, self.splits[split]))
  File "/home/yifanmai/oss/helm/src/helm/benchmark/scenarios/cleva_scenario.py", line 957, in process_dialogue_instance
    self.process_instance(
  File "/home/yifanmai/oss/helm/src/helm/benchmark/scenarios/cleva_scenario.py", line 463, in process_instance
    instance = self.converter.transform(row, self.prompt_template, split)
  File "/home/yifanmai/oss/helm/src/helm/benchmark/scenarios/cleva_scenario.py", line 70, in transform
    transformed_data = self._apply_all(copy.deepcopy(data), templates)
  File "/home/yifanmai/oss/helm/src/helm/benchmark/scenarios/cleva_scenario.py", line 203, in _apply_all
    assert isinstance(v, str)
AssertionError
```

Could you take a look please?

lyy1994 commented:
> I tried running this on the debugging model `simple/model1`. Most runs work, but some are failing inside template rendering:
> 
> Failing runs:
> 
> ```
> cleva:task=dialogue_generation,version=v1,prompt_id=0,subtask=task_oriented,model=simple/model1,data_augmentation=cleva
> cleva:task=paraphrase_identification,version=v1,prompt_id=0,subtask=short_utterance,model=simple/model1,data_augmentation=cleva
> cleva:task=paraphrase_identification,version=v1,prompt_id=0,subtask=financial_question,model=simple/model1,data_augmentation=cleva
> cleva:task=summarization,version=v1,prompt_id=0,subtask=dialogue_summarization,model=simple/model1,data_augmentation=cleva
> cleva:task=closed_book_question_answering,version=v1,prompt_id=0,subtask=medical_question_answering,model=simple/model1,data_augmentation=cleva
> cleva:task=text_classification,version=v1,prompt_id=0,subtask=news,model=simple/model1,data_augmentation=cleva
> cleva:task=text_classification,version=v1,prompt_id=0,subtask=humor,model=simple/model1,data_augmentation=cleva
> cleva:task=sentiment_analysis,version=v1,prompt_id=0,model=simple/model1,data_augmentation=cleva
> cleva:task=coreference_resolution,version=v1,prompt_id=0,model=simple/model1,data_augmentation=cleva
> cleva:task=commonsense_reasoning,version=v1,prompt_id=0,subtask=textual_entailment,model=simple/model1,data_augmentation=cleva
> cleva:task=code_synthesis,version=v1,prompt_id=0,model=simple/model1,data_augmentation=cleva
> cleva:task=toxicity_detection,version=v1,prompt_id=0,model=simple/model1,data_augmentation=cleva
> cleva:task=bias,version=v1,prompt_id=0,subtask=dialogue_gender_bias,model=simple/model1,data_augmentation=cleva
> cleva:task=bias,version=v1,prompt_id=0,subtask=dialogue_occupation_bias,model=simple/model1,data_augmentation=cleva
> cleva:task=bias,version=v1,prompt_id=0,subtask=dialogue_race_bias,model=simple/model1,data_augmentation=cleva
> cleva:task=bias,version=v1,prompt_id=0,subtask=dialogue_region_bias,model=simple/model1,data_augmentation=cleva
> cleva:task=fact_checking,version=v1,prompt_id=0,model=simple/model1,data_augmentation=cleva
> ```
> 
> Error traceback:
> 
> ```
> Traceback (most recent call last):
>   File "/home/yifanmai/oss/helm/src/helm/benchmark/runner.py", line 173, in run_all
>     self.run_one(run_spec)
>   File "/home/yifanmai/oss/helm/src/helm/benchmark/runner.py", line 221, in run_one
>     instances = scenario.get_instances()
>   File "/home/yifanmai/oss/helm/src/helm/benchmark/scenarios/cleva_scenario.py", line 942, in get_instances
>     instances.extend(self.process_dialogue_instance(row, self.splits[split]))
>   File "/home/yifanmai/oss/helm/src/helm/benchmark/scenarios/cleva_scenario.py", line 957, in process_dialogue_instance
>     self.process_instance(
>   File "/home/yifanmai/oss/helm/src/helm/benchmark/scenarios/cleva_scenario.py", line 463, in process_instance
>     instance = self.converter.transform(row, self.prompt_template, split)
>   File "/home/yifanmai/oss/helm/src/helm/benchmark/scenarios/cleva_scenario.py", line 70, in transform
>     transformed_data = self._apply_all(copy.deepcopy(data), templates)
>   File "/home/yifanmai/oss/helm/src/helm/benchmark/scenarios/cleva_scenario.py", line 203, in _apply_all
>     assert isinstance(v, str)
> AssertionError
> ```
> 
> Could you take a look please?

Thanks for letting us know about these failed runs. We have fixed these issues that mostly come from inappropriate assertions and now they work just fine on our side. We notice that fairness perturbations do not work with the code scenarios, including ours and [HumanEval](https://github.com/stanford-crfm/helm/blob/main/src/helm/benchmark/scenarios/code_scenario.py). An example config we run into error is `code:model=simple/model1,dataset=humaneval,data_augmentation=fairness`.

Traceback:
```
Traceback (most recent call last):
  File "/home/ubuntu/liyanyang/helm/src/helm/benchmark/runner.py", line 173, in run_all
    self.run_one(run_spec)
  File "/home/ubuntu/liyanyang/helm/src/helm/benchmark/runner.py", line 260, in run_one
    metric_result: MetricResult = metric.evaluate(
  File "/home/ubuntu/liyanyang/helm/src/helm/benchmark/metrics/metric.py", line 148, in evaluate
    results: List[List[Stat]] = parallel_map(
  File "/home/ubuntu/liyanyang/helm/src/helm/common/general.py", line 232, in parallel_map
    results = list(tqdm(executor.map(process, items), total=len(items), disable=None))
  File "/mnt/data/huangyongfeng/AI4Protein/anaconda3/envs/crfm-helm/lib/python3.8/site-packages/tqdm/std.py", line 1195, in __iter__
    for obj in iterable:
  File "/mnt/data/huangyongfeng/AI4Protein/anaconda3/envs/crfm-helm/lib/python3.8/concurrent/futures/_base.py", line 619, in result_iterator
    yield fs.pop().result()
  File "/mnt/data/huangyongfeng/AI4Protein/anaconda3/envs/crfm-helm/lib/python3.8/concurrent/futures/_base.py", line 437, in result
    return self.__get_result()
  File "/mnt/data/huangyongfeng/AI4Protein/anaconda3/envs/crfm-helm/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/mnt/data/huangyongfeng/AI4Protein/anaconda3/envs/crfm-helm/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/ubuntu/liyanyang/helm/src/helm/benchmark/metrics/metric.py", line 77, in process
    self.metric.evaluate_generation(
  File "/home/ubuntu/liyanyang/helm/src/helm/benchmark/metrics/basic_metrics.py", line 745, in evaluate_generation
    stats.extend(self.compute_reference_metrics(adapter_spec, request_state, metric_service))
  File "/home/ubuntu/liyanyang/helm/src/helm/benchmark/metrics/basic_metrics.py", line 551, in compute_reference_metrics
    stats.extend(compute_metrics_helper(MetricName(metric_name), metric_fn_mapping[metric_name]))
  File "/home/ubuntu/liyanyang/helm/src/helm/benchmark/metrics/basic_metrics.py", line 479, in compute_metrics_helper
    score_1 = max(score_func((gold.output.text, gold.test_cases), preds[0]) for gold in code_golds)
  File "/home/ubuntu/liyanyang/helm/src/helm/benchmark/metrics/basic_metrics.py", line 479, in <genexpr>
    score_1 = max(score_func((gold.output.text, gold.test_cases), preds[0]) for gold in code_golds)
  File "/home/ubuntu/liyanyang/helm/src/helm/benchmark/metrics/basic_metrics.py", line 366, in code_eval
    assert gold[1] is not None  # gold[1]["canonical_solution"]
AssertionError
```

Currently we use `cleva_robustness` for code scenarios in `run_specs_cleva_v1.conf` to circumvent this issue.

yifanmai commented on src/helm/benchmark/metrics/machine_translation_metrics.py:
```diff
@@ -39,3 +40,50 @@ def evaluate_instances(self, request_states: List[RequestState]) -> List[Stat]:
+
+        bleu_score = bleu.corpus_score(sys, refs).score
+
+        return [Stat(MetricName("bleu")).add(bleu_score)]
```
It would be good to call this `cleva_machine_translation_bleu` or `cleva_bleu` or `machine_translation_bleu`, since there's already a number of different metrics with names that are variations on `bleu` (in this PR and in the main branch). So this would help disambiguate.

yifanmai commented on src/helm/benchmark/run_specs.py:
```diff
@@ -2243,6 +2377,118 @@ def get_anthropic_hh_rlhf_spec(num_respondents: int, subset: str) -> RunSpec:
+    class_name_prefix = "".join([word.capitalize() for word in task.split("_")])
+    scenario_spec = ScenarioSpec(
+        class_name=f"helm.benchmark.scenarios.cleva_scenario.CLEVA{class_name_prefix}Scenario",
+        args={"version": version, "subtask": subtask, "prompt_template": prompt_template},
```
Could you make the arg `"prompt_id": prompt_id` instead of `"prompt_template": prompt_template`, and then fetch the template in `CLEVAScenario.__init__()`?

`ScenarioSpec.args` can't be too long because we use it to generate a file name, and `prompt_template` can be very long.

yifanmai commented on src/helm/benchmark/scenarios/cleva_scenario.py:
```diff
@@ -0,0 +1,1608 @@
+        self,
+        version: str,
+        subtask: str,
+        prompt_template: Dict[str, Any],
```
Replace ` prompt_template: Dict[str, Any]` with `prompt_id: int` and then inside the `__init__`() body do:

```python
self.prompt_template, _ = CLEVAScenario.get_prompt_setting(task, subtask, version, prompt_id)
```

yifanmai reviewed (changes requested):
Thank you, I was able to run it locally now! We're very close to merging. I tried running it end-to-end again and found a few issues; we should be good after this round.

yifanmai commented:
Also, please review #1845 for the schema changes... for some reason GitHub is not letting me add you as a reviewer.

lyy1994 commented:
> Thank you, I was able to run it locally now! We're very close to merging. I tried running it end-to-end again and found a few issues; we should be good after this round.

Thanks for your suggestions! We have made the necessary changes according to your comments and tested `run_specs_cleva_v1.conf` with `simple/model1` on our side successfully.

> Also, please review #1845 for the schema changes... for some reason GitHub is not letting me add you as a reviewer.

We have reviewed #1845 and @Jianqiao-Zhao left a comment that we have no problem with that template. We managed to open a new PR soon to fill `schema.yaml` after this PR merged. Thanks for your guidance in the past few weeks!

yifanmai commented:
Sorry, it looks like I just caused some merge conflicts by landing #1844. Could you resolve these please?

The required changes should be to rename your `requirements-freeze.txt` file to `requirements.txt` (use your latest contents), deconflict the requirements extras lists in extras.cfg, and redo the imports in run.py

Thank you!

lyy1994 commented:
> Sorry, it looks like I just caused some merge conflicts by landing #1844. Could you resolve these please?
> 
> The required changes should be to rename your `requirements-freeze.txt` file to `requirements.txt` (use your latest contents), deconflict the requirements extras lists in extras.cfg, and redo the imports in run.py
> 
> Thank you!

I think we have resolved the conflicts, but Github still reports some weird conflict cases in `requirements.txt` and `setup.cfg`... Another issue is that we only handle conflicts in `src/helm/benchmark/run.py`, `requirements.txt` and `setup.cfg`. But #1844 also changes other files, making the Github action fail the test. Please let us know if we need to take more action.

yifanmai commented:
The easiest way to do this would be through the web editor.

If that doesn't work, you can do the following on the command line (assuming your local main branch is set to track `lyy1994:main`):

```bash
git checkout main
git pull https://github.com/stanford-crfm/helm.git main
# TODO: Manually edit setup.cfg to resolve the conflict
# i.e. add the CLEVA dependencies
git add setup.cfg
git commit
git log
# TODO: Check that the latest commit is called
# "Merge branch 'main' of https://github.com/stanford-crfm/helm into main"
git push
```

yifanmai commented:
I forgot to say: to use the web editor, click the "Resolve conflicts" button next to "This branch has conflicts that must be resolved".

Sorry again for the trouble.

lyy1994 commented:
> The easiest way to do this would be through the web editor.
> 
> If that doesn't work, you can do the following on the command line (assuming your local main branch is set to track `lyy1994:main`):
> 
> ```shell
> git checkout main
> git pull https://github.com/stanford-crfm/helm.git main
> # TODO: Manually edit setup.cfg to resolve the conflict
> # i.e. add the CLEVA dependencies
> git add setup.cfg
> git commit
> git log
> # TODO: Check that the latest commit is called
> # "Merge branch 'main' of https://github.com/stanford-crfm/helm into main"
> git push
> ```

Thanks for your detailed comment! The conflicts should all be resolved now.
